Docs: Update preferred_practices.md

patrinoua opened the pull request:
The type of this PR is: **docs**

This PR updates the phrasing and removes some old parts of the documentation. 

preferred_practices.md is there to give us a broad overview of the history and how we prefere to do things.

@olerichter00 could you maybe contribute on the **relay hooks** part?

@gkartalis could you contribute on the **testing** part? 

We don't need to fix everything under this PR, but this could be a start to make sure we keep our current practices up to date.
 
<!-- Bugfix/Feature/Enhancement/Documentation -->

<!-- If applicable, write the Jira ticket number in square brackets e.g. [PROJECT-XXXX]
    The Jira integration will turn it into a clickable link for you. -->

### Description

### PR Checklist (tick all before merging)

<!-- 💡 This checklist is experimental. CX warmly welcomes any feedback about the list or how it impacts your workflow -->

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [x] I have tested my changes on **iOS** and **Android**.
- [x] I have added tests/stories for my changes, or my changes don't require testing/stories, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have added a feature flag, or my changes don't require a feature flag. ([How do I add one?](https://github.com/artsy/eigen/blob/master/docs/developing_a_feature.md))
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/master/docs/adding_state_migrations.md))
- [x] I have added a changelog entry below or my changes do not require one.

### To the reviewers 👀

- [ ] I would like at least one of the reviewers to run this PR on the simulator or device.

<details><summary>Changelog updates</summary>

### Changelog updates

<!-- 📝 Please fill out at least one of these sections. -->
ⓘ 'User-facing' changes will be published as release notes.
<!-- ⌫ Feel free to remove sections that don't apply. -->
<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
<!-- 📖 eg. `Enable lotsByFollowedArtists - john` or `Fix phone input misalignment - mary`. -->
<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->

#### Cross-platform user-facing changes

- Rename modal title - patrinoua

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

<!-- end_changelog_updates -->

</details>

ArtsyOpenSource commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



### This PR contains the following changes:

- Cross-platform user-facing changes (Rename modal title - patrinoua)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against bbc4a2903d5b592ea6b3f534ac3dbf23789d2c2c
</p>

admbtlr commented on docs/preferred_practices.md:
```diff
@@ -26,39 +21,26 @@ Links should point to specific commits, and not a branch (in case the branch or
-
-New features should be built in React Native. The React Native runtime currently requires an existing user ID and access token to be loaded, and sign up/in is still handled in Native code.
+The app was initially written in Objective-C and Swift and React Native was added in 2016. Some parts of the app are also written with Kotlin.
+New features should be built in React Native.
```
Something like `Swift and Kotlin should only be used for "glue" code, or native functionality that is impossible to implement with React Native. Objective-C should be avoided.`

brainbicycle replied:
minor delta from RFC: https://github.com/artsy/README/pull/426 Objective-C can still be used for the thin bridging code, Swift for major feature work.

patrinoua replied:
@brainbicycle Thanks! 
Since you are more familiar with this topic, could you perhaps help me with some more context and the right phrasing here?

How about something like this:

- **React Native** By default we use React Native for new features. 
- **Kotlin**  should only be used for "glue" code, or native functionality that is impossible to implement with React Native. 
- **Objective-C** can be used for the thin bridging code (personal note: i am not sure what that means) and 
- **Swift** for major feature work that can't be done in React Native (such as: ___).

brainbicycle replied:
Sure! I think it should be something like this:

- **React Native** By default we use React Native for new features.
- **Objective-C**, Java can be used for bridging code to react native (this is referring to native modules that need to talk to javascript, more info in the react native docs here: https://reactnative.dev/docs/native-modules-ios)
- **Swift, Kotlin** for native functionality that can't be done in React Native (such as: an iOS Widget or a Push Notification Extension).

brainbicycle replied:
let me know if you have any questions on that

admbtlr commented on docs/preferred_practices.md:
```diff
@@ -120,30 +103,32 @@ However, if we have a `Button` folder which exports only one button component, w
+    We recently also started using relay-hooks and will soon link documents to that.
 
-### Prefer Relay containers (Higher Order Components) over Hooks
+### Prefer Relay containers (Higher Order Components) over Hooks // !! is this still valid?
```
We should prefer relay hooks for all new implementations. Refactoring old containers to use hooks is encouraged.

admbtlr commented on docs/preferred_practices.md:
```diff
@@ -120,30 +103,32 @@ However, if we have a `Button` folder which exports only one button component, w
-### Prefer Relay containers (Higher Order Components) over Hooks
+### Prefer Relay containers (Higher Order Components) over Hooks // !! is this still valid?
 
 We have a preference for Relay containers due to `relay-hooks` hooks not being compatible with Relay containers which represent the majority of our components using Relay.
```
Pretty sure this can be removed - right @olerichter00 ?

admbtlr commented on docs/preferred_practices.md:
```diff
@@ -120,30 +103,32 @@ However, if we have a `Button` folder which exports only one button component, w
 
 We have a preference for Relay containers due to `relay-hooks` hooks not being compatible with Relay containers which represent the majority of our components using Relay.
 
 - [Relay Container approach](https://github.com/artsy/eigen/blob/21fbf9e24eaa281f3e16609da5d38a9fb62a5449/src/lib/Scenes/MyAccount/MyAccount.tsx#L70)
```
and this should be removed and replaced with a canonical relay hooks-based component

admbtlr commented:
Thanks @patrinoua , this is a great start!

brainbicycle commented:
Thanks for taking initiative on this! I can volunteer to update the navigation section which is way out of date at this point.

pvinis commented on docs/preferred_practices.md:
```diff
@@ -120,30 +103,32 @@ However, if we have a `Button` folder which exports only one button component, w
 
 ### styled-system / styled-components
 
+// !! is this still valid?
```
i feel like the following two list items could go away and be replaced by something like `We use [palette](src/palette) which uses styled-system under the hood`.

olerichter00 commented:
@patrinoua Thanks a lot, this is great work! 🌟 

I'll create another PR and revise sections/docs about Relay according to this [RFC](https://github.com/artsy/eigen/issues/6080) about using Relay Hooks in favor of the container-based APIs.

olerichter00 commented:
I've updated the sections about Relay in https://github.com/artsy/eigen/pull/6287

pvinis commented on docs/preferred_practices.md:
```diff
@@ -1,64 +1,54 @@
-- Deep-link and notification handling (via SwitchBoard).
-- Analytics for Native UI.
-- Initializing the React Native runtime.
+Check out our list of [examples](#../../../EXAMPLES.md) and [hacks](#../../../HACKS.md)
```
```suggestion
Check out our list of [examples](/EXAMPLES.md) and [hacks](/HACKS.md)
```
I think this should make it link nicely

brainbicycle commented on docs/preferred_practices.md:
```diff
@@ -120,64 +87,56 @@ However, if we have a `Button` folder which exports only one button component, w
+### Navigation
 
-#### Follow the tracking docs and examples
+We use `react-navigation` for navigating between screens, navigate function and new routes.
```
```suggestion
We use `react-navigation` for navigating between screens. For adding a screen that corresponds to a page on artsy.net use the `navigate(<route-name>)` function and add a new route.
```

MounirDhahri replied:
@patrinoua we don't actually use react-navigation for navigation in iOS. we use instead the Native Switchboard there. I would probably mention that and then do as you did by referring users to the add a new route docs

patrinoua replied:
@brainbicycle @MounirDhahri can you please have a look at this and come up with the right wording for this section and create a PR that we can add here?

brainbicycle reviewed (changes requested):
minor wording change request on navigation, otherwise looks great to me!

patrinoua commented on docs/preferred_practices.md:
```diff
@@ -120,30 +109,32 @@ However, if we have a `Button` folder which exports only one button component, w
 `Note:` Updating capitalisation on folders can cause issues in git and locally so please refrain from renaming existing folders until we come up with a strategy about this. (TODO)
 
-### Use Relay for Network Requests
+### Relay
```
Thanks Brian, I added "We use relay for network requests." to make it even more clear, while keeping it friendly 🤣, is it ok like this?

pvinis commented:
Do you wanna rename the file in this PR too?

MounirDhahri commented on docs/preferred_practices.md:
```diff
@@ -120,64 +87,56 @@ However, if we have a `Button` folder which exports only one button component, w
+### Analytics and tracking
 
-- [Making network requests outside of Relay](https://github.com/artsy/emission/blob/019a106517b31cebfb1c5293891215cc7ebf7a4d/src/app/Components/Consignments/Screens/Overview.tsx#L135-L150)
+Implementing analytics and tracking [here](./analytics_and_tracking.md).
```
In React-native, we use `react-tracking` as a wrapper for the tracking events we send to Segment. You can read more about the implementation [here](./analytics_and_tracking.md).

patrinoua commented on docs/preferred_practices.md:
```diff
@@ -1,80 +1,47 @@
 
-As we have built our app in React Native, our understanding of how to build software has evolved. As our understanding grows, newer code uses newer techniques. Older code is often left un-updated. It can be difficult to orient oneself around what the current preferred practices are.
+This is a living document, expected to be updated regularly, with a broad overview of the history and how we prefere to do things on eigen.
+Here you can find links to the tools we use, examples, pull requests with interesting discussions & blog posts.
 
-This document is a map. Not of Eigen at a specific time, but a map of how we got here and where we want to go next. This is a living document, expected to be updated regularly, of links to:
-
-- Example code.
-- Pull requests with interesting discussions.
-- Conversations on Slack.
-- Blog posts.
-
-Links should point to specific commits, and not a branch (in case the branch or file is deleted, these links should always work). But it's possible that a file is outdated, that our understanding has moved on since it was linked to; in that case, please update this document.
```
I thought I'd make it a bit shorter and less "instructive", as anyway we have quite some reviewing before we add something here.
So I removed the instruction about the edge case of "do not add a link to a branch, but to a commit".

Some parts of the history are also mentioned later.

In general I thought the voice before sounded like "we are struggling a bit" and tried to make it less of that, while still conveying the same information. 

Would you like to bring back some specific thing that is missing or revert all of it?

MounirDhahri replied:
I personally find some of those tips useful but wanted to check with you to understand your point of view. I am fine with removing them definitely

MounirDhahri commented on docs/preferred_practices.md:
```diff
@@ -1,80 +1,47 @@
-Objective-C and Swift (sometimes called "Native" code) are responsible for the following parts of the app:
+Check out our lists of [examples](#../../../EXAMPLES.md) and [hacks](#../../../HACKS.md).
 
-- Sign up/in flow ("onboarding").
-- Live Auctions Integration (LAI) view controller and networking.
-- The Auction view controller.
-- The SwitchBoard (see "SwitchBoard" section below) to navigate between view controllers.
-- The top-level tab bar, and each tab's navigation controller.
-- Deep-link and notification handling (via SwitchBoard).
-- Analytics for Native UI.
-- Initializing the React Native runtime.
```
I think here we are describing the parts of the app that are still being handled in native instead of react-native on iOS and this is the only place where they're being described

brainbicycle replied:
@admbtlr added a section with this so I think this can be marked resolved!

MounirDhahri reviewed (dismissed):
Thanks for leading this effort Angela and updating those really old docs. I only have one comment left unresolved about mentioning the parts that are still being handled in native, but it's not blocking.
Thanks again

patrinoua commented:
@MounirDhahri , @brainbicycle Thanks once again for your feedback and help on meticulously updating our docs. 🙏 
Apart from the benefit this will have for eigen and our newly joining developers it also helps me get a better understanding of how we get things done.

With my latest commit
- I've updated the parts that have to do with navigation and added a section on Android and iOS parts.
- I've added a Miscellaneous section and as a subsection "Parts of the app that are still being handled in native code (Objective-C and Swift) instead of react-native on iOS" with the list of old components still being used.

--- 

One thing I'm not 100% on board with and would like to discuss a bit further is adding links to commits:

I consider this to be an edge case because I'm thinking that if we point to a file as a best practice for example, and then the file gets a more recent update, the more recent update is also going to have an even better practice.

While updating these docs, I've added a couple of links to files but not PRs/branches/commits. 

There are a couple of links that point to commits instead of files left under testing. And I can't say I'd feel 100% good about using those without asking, as I'd wonder if they are up to date, since some of them are from 2020 and a lot has changed since then.

So I think that adding a commit to a file that is on the main branch is helping us from keeping things that are outdated. 
Of course if the file gets removed then ideally we also eventually remove it from our docs. 
But I guess this is a far less the case, than committing more things to a relatively active part of the app / component.

This is why I think the standard practice should be adding links to files on main and not just a specific commit when it comes to files.

In general I am not in favour of pointing to PRs/branches but is there a good case where we do that? Do we really need it as part of our preferred_practices ?

I'd love to also see your thoughts about that 🙏

MounirDhahri commented:
> One thing I'm not 100% on board with and would like to discuss a bit further is adding links to commits:
> 
> I consider this to be an edge case because I'm thinking that if we point to a file as a best practice for example, and then the file gets a more recent update, the more recent update is also going to have an even better practice.

The issue with linking files is that they get bigger, get more complicated and someone else can for example update the file and might mistakenly use a non-recommended practice. Files location can change as well, for example after the latest `src/app` rename, all our links to files would have broke if we were using links to files.  
The downside of linking to commits IMO is that the file that we need to link to inside a commit, needs to be as good as we can get it since it won't get updates later.
In both cases, as is the case always for docs, no matter how well it's written, since we're making updates to our codebase, we will need always to update our docs.

patrinoua commented:
> > One thing I'm not 100% on board with and would like to discuss a bit further is adding links to commits:
> > I consider this to be an edge case because I'm thinking that if we point to a file as a best practice for example, and then the file gets a more recent update, the more recent update is also going to have an even better practice.
> 
> The issue with linking files is that they get bigger, get more complicated and someone else can for example update the file and might mistakenly use a non-recommended practice. Files location can change as well, for example after the latest `src/app` rename, all our links to files would have broke if we were using links to files. The downside of linking to commits IMO is that the file that we need to link to inside a commit, needs to be as good as we can get it since it won't get updates later. In both cases, as is the case always for docs, no matter how well it's written, since we're making updates to our codebase, we will need always to update our docs.

@MounirDhahri if we have an agreed way of doing things the non recommended practice should get caught on review. 
This is the mechanism for maintaining a good code base and why we need to have agreed best practices...

brainbicycle commented:
> > > One thing I'm not 100% on board with and would like to discuss a bit further is adding links to commits:
> > > I consider this to be an edge case because I'm thinking that if we point to a file as a best practice for example, and then the file gets a more recent update, the more recent update is also going to have an even better practice.
> > 
> > 
> > The issue with linking files is that they get bigger, get more complicated and someone else can for example update the file and might mistakenly use a non-recommended practice. Files location can change as well, for example after the latest `src/app` rename, all our links to files would have broke if we were using links to files. The downside of linking to commits IMO is that the file that we need to link to inside a commit, needs to be as good as we can get it since it won't get updates later. In both cases, as is the case always for docs, no matter how well it's written, since we're making updates to our codebase, we will need always to update our docs.
> 
> @MounirDhahri if we have an agreed way of doing things the non recommended practice should get caught on review. This is the mechanism for maintaining a good code base and why we need to have agreed best practices...

I don't really agree with this, files move and change all the time in a code base and linking to a commit ensures we are linking to exactly the practice we are trying to showcase at a point in time. Say we add a non-commit link to an example test file for testing relay responses but the associated component gets deleted or renamed a few months down the line, now your link is dead. I don't really think we can reasonably expect people to be checking these links every time new code gets added or removed from the codebase. Similarly maybe the data fetching got moved to another component so now the test doesn't show the preferred practice we were trying to showcase at all anymore but might still adhere to best practices otherwise.

patrinoua commented:
@brainbicycle @MounirDhahri thanks for elaborating more on the commit to link vs file topic. 🙏
I see your point and already on Friday added a commit bringing that instruction back, but probably forgot to push it! I've added today though. [link](https://github.com/artsy/eigen/pull/6243/commits/d54a75039e7c47a6c4c01b7127da2392b512e75c) 
Shall we have another look and if everything is ok merge?

MounirDhahri reviewed (approved):
Great, this sounds good to me. This is a big step in refreshing Eigen docs. Thanks @patrinoua for leading this big effort

brainbicycle reviewed (approved):
Woo! Thanks a ton for doing this and pushing through all our comments! This is much better!

patrinoua commented:
🎉 🥳 🎉 We meeerged! 🎉 🥳 🎉

Next I'll make a small PR renaming preferred_practices to best_practices as was discussed in #mobile-practice.

Thank you all so much for contributing to this!
